APE18.rst

```diff
@@ -0,0 +1,69 @@
+Adopt NEP 29 for CPython and Numpy Version Support
+--------------------------------------------------
+
+author: Stuart Mumford and Thomas Robitaille
+
+date-created: 2020 October 16
+
+date-last-revised: 2020 October 16
+
+type: Standard Track
+
+status: Discussion
+
+
+Abstract
+--------
+
+After the release of astropy 4.0 APE 10 no longer specifies the supported versions of CPython.
+Since APE 10 was written the wider scientific Python community have written and largely adopted `NEP 29 <https://numpy.org/neps/nep-0029-deprecation_policy.html>`__ as a policy for supported versions of both CPython and Numpy.
+This APE proposes that Astropy follow the projects who have already adopted this policy.
+
+
+Detailed description
+--------------------
+
+This APE fully adopts NEP 29 for both CPython and Numpy versions, it supersedes APE 10 wherever there is a conflict between that APE and this APE or NEP 29.
+
+While this APE will not replicate the whole of NEP 29, and NEP 29 should be considered the canonical source for the support policy, the core scheme to be adopted is as follows:
+
+astropy will support:
```

Comment:
The core astropy package as well as other coordinated and infrastructure packages will support

Reply:
This is perhaps something different from APE 10, that this would be project wide

Reply:
I would not force infra and coordinated to follow this, e.g. they currently usually support more, especially the infra packages.

Reply:
So I would just go that they can't drop more than what is suggested in NEP29, but they could still support older versions of they want to

Reply:
Yeah, not sure how we can enforce this project-wide in a realistic manner. We barely have enough people to maintain the core as it is.

Reply:
I have scoped this to core, not sure we really need to do anything else here?

Reply:
@pllim - and my thinking is: not sure why we _want_ to enforce this project wide.

Reply:
I think it's important that none of the infra/coordinated drop things sooner and switch to python 3.10+ now, but if they have a reason to still support 3.4, then why not (usually it only means that they didn't run into an issue and everything still works, while dropping is more work)

Reply:
Maybe we should say the "rule" applies to core, but coordinated&infrastructure packages will "aim" to follow this unless there's a reason not to?  I.e. just say for everything else it's a guideline.

I think the virtue of that is that it gives some guidance for packages that are sort of like "we don't have a rule because we haven't really thought that deeply about it".  Or alternatively/additionally, it provides an excuse to ask institutional players to help us bring things "up to standard" (I'm thinking here about how the py 3.x transition provided some people at institutions a way to fund additional work by existing Astropy devs)

Reply:
many of the infra packages didn't really have a reason to drop old versions for a while, and I fail to see why it would be different in the future. As I said above, in many cases dropping a version is more work than keeping it, thus the "aim" is problematic in practice. And frankly, what is important for the ecosystem is that all coordinated and infra support at least the NEP29, that there is no situation when something gets dropped sooner than this schedule.

Reply:
Agree we can make it a rule for core and recommendation for other parts of the project?

Reply:
>  if they have a reason to still support 3.4, then why not 

Just thinking about #10888 that I just put in, which would require 3.7+ for astropy to work properly. If a coordinated package tries to force using Python 3.4 + astropy 4.2 then there will be problems. Not sure if that is credible.

Reply:
I'm not sure what you mean credible here, it's always up to the package manager to resolve the version requirements, I don't see any force being involved here.

E.g. on 3.4, it could be possible to install the latest version of a coordinated package, and naturally, it would come with an older astropy version. It's down to the CI of the coordinated package to make sure they are compatible with their minimum supported version combination, after all the user should know best why they are stuck on an older python/numpy combo.

(3.4 is probably a bad example, but a real life example, with astroquery we had to stay python2.7 compatible much much longer than astropy core as it took long for casa to move off it. We, the astroquery maintainers decided that it's something we were willing to do, and it was totally independent of the version support policy of core that was decided separately. )